proposals/XXXX-extend-property-wrappers-to-function-and-closure-parameters.md

```diff
@@ -0,0 +1,473 @@
+
+This limitation in expressivity is emphasized by the fact that property wrappers were originally sought out to abstract away such patterns.  As a result, elegant APIs are undermined by this limitation. Not only is this limiting users by forcing them to carefully read documentation, which may not cover a specific use case, to make sure no invariants have been violated, but it also limits API authors in what they can create. That is, API authors can't use property-wrapper types in closure parameters nor can code be seperated into functions that accept property wrapper syntax:
+
+```swift
+extension Percentage {
+
+  func modify(
+    inSeconds seconds: Int,
+    block: @escaping (Clamped<Int>) -> Void
+  ) { ... }
+  
+}
+
+
+let myPercentage = Percentage(percent: 50)
+
+myPercentage
+  .modify(inSeconds: 3) { percent in
+    percent.wrappedValue = 100
+    //    ^~~~~~~~~~~~ 
+    // Again, we have to access
+    // count through 'wrappedValue'.
+  }
+```
+
+In fact, establishing custom behavior on closure parameters is really powerful. For example, if such a feature were supported, it could be used in conjunction with [Function Builders](https://github.com/apple/swift-evolution/blob/master/proposals/0289-function-builders.md) to expose data managed by a 'component' type. For instance, in SwiftUI [`ForEach`](https://developer.apple.com/documentation/swiftui/foreach) could leverage this feature to expose the mutable state of its data source to its 'content' closure. Thus, instead of manually mutating the data source, as is done here:
+
+```swift
+struct MyView: View {
+
+  // A simple Shopping Item that includes
+  // a 'quantity' and a 'name' property.
+  @State 
+  private var shoppingItems: [Item]
+
+
+  var body: some View {
+    ForEach(0 ..< shoppingItems.count) { index in
+  
+      Text(shoppingItems[index].name)
+        .onTapGesture {
+          // We increase the item's quantity 
+          // when the user taps the item. 
+          // Unfortunately, to mutate the item
+          // we have to manually index our
+          // data source.
+          shoppingItems[index].quanity += 1
+        }
+      
+    }
+  }
+  
+}
+```
+
+we would – with an appropriate initializer – be able to simplify the above code, reducing boilerplate as a result:
+
+```swift
+struct MyView: View {
+
+  @State 
+  private var shoppingItems: [Item]
+
+
+  var body: some View {
+    ForEach($shoppingItems) { @Binding shoppingItem in
+    
+      Text(shoppingItem.name)
+        .onTapGesture {
+          shoppingItem.quanity += 1
```

Comment:
Can we add a `TextField` or something else that requires a binding via `$shoppingItem` to this example to show that using an `inout Item` parameter would not really solve the issue?